app/core/secure_prompt_manager.py

```
@@ -0,0 +1,346 @@
+            if key in self.placeholders:
+                sanitized_data[key] = sanitize_text(str(value))
+            else:
+                # Skip unknown keys to prevent injection
+                logger.warning(f"Unknown placeholder key: {key}")
+                continue
+        
+        # Use safe string formatting with explicit placeholders
+        try:
+            return self.template.format(**sanitized_data)
+        except KeyError as e:
+            raise ValueError(f"Template rendering error: {e}")
```

Ruff (0.12.2)

Within an `except` clause, raise exceptions with `raise ... from err` or `raise ... from None` to distinguish them from errors in exception handling

(B904)